File: 572.subtree-of-another-tree.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if root is None:
            return False
        if root.val == subRoot.val and self.dfs(root, subRoot):
            return True
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

    def dfs(self, p, q):
        if p is None and q is None:
            return True
        if p is None or q is None or p.val != q.val:
            return False
        return self.dfs(p.left, q.left) and self.dfs(p.right, q.right)


# @leet end

File: 572.subtree-of-another-tree.c
// @leet start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool dfs(struct TreeNode *p, struct TreeNode *q) {
  if (!p && !q) {
    return true;
  }
  if (!p || !q || p->val != q->val) {
    return false;
  }
  return dfs(p->left, q->left) && dfs(p->right, q->right);
}
bool isSubtree(struct TreeNode *root, struct TreeNode *subRoot) {
  if (!root) {
    return false;
  }
  if (root->val == subRoot->val && dfs(root, subRoot)) {
    return true;
  }
  return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
}
// @leet end
